cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,388 @@
+	}
+
+	// parse timeout
+	failureTimeout, err := parseTimeout(strategy.All.Timeout.Timeout)
+	if err != nil {
+		return &strategy, RolloutResult{}, err
+	}
+
+	// get all the clusters
+	totalClusters, _, totalClusterGroups := r.pdTracker.ExistingBesides([]clusterv1beta1.GroupKey{})
+	clusterToGroupKey := r.pdTracker.ClusterToGroupKey(totalClusterGroups)
+	rolloutresult := progressivePerCluster(totalClusters.UnsortedList(), clusterToGroupKey, len(totalClusters), failureTimeout, statusFunc)
+
+	return &strategy, rolloutresult, nil
+}
+
+func (r *RolloutHandler) getProgressiveClusters(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, RolloutResult, error) {
+	strategy := RolloutStrategy{Type: Progressive}
+	strategy.Progressive = rolloutStrategy.Progressive.DeepCopy()
+	if strategy.Progressive == nil {
+		strategy.Progressive = &RolloutProgressive{}
+	}
+
+	// upgrade mandatory decision groups first
+	groupKeys := decisionGroupsToGroupKeys(strategy.Progressive.MandatoryDecisionGroups.MandatoryDecisionGroups)
+	clusterGroupKeys, clusterGroups := r.pdTracker.ExistingClusterGroups(groupKeys)
+
+	rolloutresult := progressivePerGroup(clusterGroupKeys, clusterGroups, maxTimeDuration, statusFunc)
+	if len(rolloutresult.ClustersToRollout) > 0 {
+		return &strategy, rolloutresult, nil
+	}
+
+	// parse timeout
+	failureTimeout, err := parseTimeout(strategy.Progressive.Timeout.Timeout)
+	if err != nil {
+		return &strategy, RolloutResult{}, err
+	}
+
+	// calculate length
+	totalClusters, _, _ := r.pdTracker.ExistingBesides([]clusterv1beta1.GroupKey{})
+	length, err := calculateLength(strategy.Progressive.MaxConcurrency, len(totalClusters))
+	if err != nil {
+		return &strategy, RolloutResult{}, err
+	}
+
+	// upgrade the rest clusters
+	restClusters, _, restClusterGroups := r.pdTracker.ExistingBesides(clusterGroupKeys)
+	restClustersToGroupKey := r.pdTracker.ClusterToGroupKey(restClusterGroups)
+	rolloutresult = progressivePerCluster(restClusters.UnsortedList(), restClustersToGroupKey, length, failureTimeout, statusFunc)
+
+	return &strategy, rolloutresult, nil
+}
+
+func (r *RolloutHandler) getProgressivePerGroupClusters(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, RolloutResult, error) {
+	strategy := RolloutStrategy{Type: ProgressivePerGroup}
+	strategy.ProgressivePerGroup = rolloutStrategy.ProgressivePerGroup.DeepCopy()
+	if strategy.ProgressivePerGroup == nil {
+		strategy.ProgressivePerGroup = &RolloutProgressivePerGroup{}
+	}
+
+	// upgrade mandatory decision groups first
+	groupKeys := decisionGroupsToGroupKeys(strategy.ProgressivePerGroup.MandatoryDecisionGroups.MandatoryDecisionGroups)
+	clusterGroupKeys, clusterGroups := r.pdTracker.ExistingClusterGroups(groupKeys)
+
+	rolloutresult := progressivePerGroup(clusterGroupKeys, clusterGroups, maxTimeDuration, statusFunc)
+	if len(rolloutresult.ClustersToRollout) > 0 {
+		return &strategy, rolloutresult, nil
```

Comment:
Same comment as above. I was not expecting this behavior based on the description of the calling function.

Reply:
I'm not sure if I understand the question here. Is expectation is to add more description for `GetRolloutCluster`, to mention that if mandatory groups defined, will return clusters to rollout in mandatory groups, and if all the mandatory groups rollout successfully, will return the rest of the clusters need to rollout ?

Reply:
Yes, I think that would be helpful.